### PR TITLE
Fixes #7280: Add generic methods to create variables

### PR DESCRIPTION
**Pull request by peckpeck**
https://www.rudder-project.org/redmine/issues/7280

### CODE REVIEW

**Comment by peckpeck** [comment]
Original PR https://github.com/Normation/ncf/pull/234

File: tests/acceptance/30_generic_methods/variable_dict.cf [hunk 1]
```diff
@@ -0,0 +1,65 @@
+#######################################################
+#
+# Launch the /bin/date command
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_dict.cf]: again, wrong comment

File: tests/acceptance/30_generic_methods/variable_dict.cf [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      version => "1.0";
+}
+
+#######################################################
+
+bundle agent init
+{
+  vars:
+    "command_0"            string => "exit 0";
+    "canonified_command_0" string => canonify("${command_name}");
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_dict.cf]: these vars don't seem used....

File: tests/acceptance/30_generic_methods/variable_dict_from_file.cf [hunk 3]
```diff
@@ -0,0 +1,62 @@
+#######################################################
+#
+# Launch the /bin/date command
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_dict_from_file.cf]: wrong comment

File: tests/acceptance/30_generic_methods/variable_iterator.cf [hunk 4]
```diff
@@ -0,0 +1,61 @@
+#######################################################
+#
+# Launch the /bin/date command
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_iterator.cf]: wrong comment

File: tests/acceptance/30_generic_methods/variable_iterator.cf [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      version => "1.0";
+}
+
+#######################################################
+
+bundle agent init
+{
+  vars:
+    "command_0"            string => "exit 0";
+    "canonified_command_0" string => canonify("${command_name}");
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_iterator.cf]: these values are not used

File: tests/acceptance/30_generic_methods/variable_iterator_from_file.cf [hunk 6]
```diff
@@ -0,0 +1,58 @@
+#######################################################
+#
+# Launch the /bin/date command
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_iterator_from_file.cf]: same

File: tests/acceptance/30_generic_methods/variable_iterator.cf [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+#######################################################
+
+bundle agent check
+{
+  classes:
+    
+    "ok_1" expression => "variable_iterator_var1_kept.!variable_iterator_var1_repaired.!variable_iterator_var1_error";
+    "ok_2" expression => "variable_iterator_var2_kept.!variable_iterator_var2_repaired.!variable_iterator_var2_error";
+    "ok_length1" expression => strcmp(length("prefix.var1"), "3");
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_iterator.cf]: this test will work only with CFEngine 3.6
Is there something that allows to skip these test when using a 3.5 agent ?

File: tests/acceptance/30_generic_methods/variable_iterator_from_file.cf [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+#######################################################
+
+bundle agent check
+{
+  classes:
+    
+    "ok_1" expression => "variable_iterator_from_file_var1_kept.!variable_iterator_from_file_var1_repaired.!variable_iterator_from_file_var1_error";
+    "ok_length1" expression => strcmp(length("prefix.var1"), "3");
+    "ok_3rd_1" expression  => strcmp(nth("prefix.var1", "2"), "lines");
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_iterator_from_file.cf]: this test will work only with CFEngine 3.6
Is there something that allows to skip these test when using a 3.5 agent ?

File: tests/acceptance/30_generic_methods/variable_string.cf [hunk 9]
```diff
@@ -0,0 +1,58 @@
+#######################################################
+#
+# Launch the /bin/date command
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_string.cf]: wrong comment

File: tests/acceptance/30_generic_methods/variable_string.cf [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    "prefix" string => "prefix";
+}
+
+#######################################################
+
+bundle agent test
+{
+  methods:
+    "ph1" usebundle => variable_string("${init.prefix}", "var1", "${init.var}");
+    "ph2" usebundle => variable_string("prefix", "var2", "value");
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_string.cf]: maybe testing with something else than prefix would be better...

File: tests/acceptance/30_generic_methods/variable_string_from_from_file.cf [hunk 11]
```diff
@@ -0,0 +1,57 @@
+#######################################################
+#
+# Launch the /bin/date command
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_string_from_from_file.cf]: wrong comment

File: tree/30_generic_methods/variable_dict.cf [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Variable dict
```
Comment (ncharles) [inline tree/30_generic_methods/variable_dict.cf]: this should have the tag saying its only 3.6 or more, for the list_compatible_inputs

File: tests/acceptance/30_generic_methods/variable_iterator.cf (hunk 7 above)
Comment (peckpeck) [inline tests/acceptance/30_generic_methods/variable_iterator.cf]: I don't think this is possible. Which function will fail ?

File: tests/acceptance/30_generic_methods/variable_string.cf (hunk 10 above)
Comment (peckpeck) [inline tests/acceptance/30_generic_methods/variable_string.cf]: What would you recommend ?

Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_string.cf]: test ?
I just want to avoid colision with ${init.prefix}

**Comment by peckpeck** [comment]
PR updated

File: tests/acceptance/30_generic_methods/variable_iterator.cf (hunk 7 above)
Comment (ncharles) [inline tests/acceptance/30_generic_methods/variable_iterator.cf]: ha, i thought length was introduced in 3.6, but actually it exists in 3.5

**Comment by ncharles** [comment]
could you simply change the prefix value, and then merge ?
thank you !

**Comment by peckpeck** [comment]
PR updated

**Comment by peckpeck** [comment]
Self merging

**Pull request merged**